cpp/tests/utilities/identify_stream_usage/identify_stream_usage.cpp

```diff
@@ -0,0 +1,329 @@
+/*
+ * Copyright (c) 2022, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <cxxabi.h>
+#include <dlfcn.h>
+#include <execinfo.h>
+#include <iostream>
+#include <stdexcept>
+#include <unordered_map>
+
+#include <cuda_runtime.h>
+
+/**
+ * @brief Print a backtrace and raise and error if stream is a default stream.
+ */
+void check_stream_and_error(cudaStream_t stream)
+{
+  // We explicitly list the possibilities rather than using
+  // `cudf::get_default_stream().value()` for two reasons:
+  // 1. There is no guarantee that `thrust::device` and the default value of
+  //    `cudf::get_default_stream().value()` are actually the same. At present,
+  //    the former is `cudaStreamLegacy` while the latter is 0.
+  // 2. Using the cudf default stream would require linking against cudf, which
+  //    adds unnecessary complexity to the build process (especially in CI)
+  //    when this simple approach is sufficient.
+  if (stream == static_cast<cudaStream_t>(0) || (stream == cudaStreamLegacy) ||
+      (stream == cudaStreamPerThread)) {
+#ifdef __GNUC__
```

Comment:
That is a good question. My godbolt investigations indicated that the answer is no, at least for clang, and since we don't actually support any other compilers for cudf at this point it probably isn't worth thinking about this too much. I only included this guard because I saw that the demangling functions are implementation-defined and I have not tested any other compilers.